src/security/PeerConnector.cc

```diff
@@ -309,14 +309,16 @@ Security::CertErrors *
 Security::PeerConnector::sslCrtvdCheckForErrors(Ssl::CertValidationResponse const &resp, Ssl::ErrorDetail *& errDetails)
 {
     ACLFilledChecklist *check = NULL;
+    Security::SessionPointer session(fd_table[serverConnection()->fd].ssl);
+
     if (acl_access *acl = ::Config.ssl_client.cert_error) {
         check = new ACLFilledChecklist(acl, request.getRaw(), dash_str);
         check->al = al;
         check->syncAle(request.getRaw(), nullptr);
+        check->serverCert.resetWithoutLocking(SSL_get_peer_certificate(session.get()));
```

Comment:
The above code implies that a developer has to remember to consider setting serverCert field before each ACLFilledChecklist check. Why is this or similar serverCert-setting logic _not_ located inside ACLFilledChecklist constructor (or equivalent) instead of burdening individual ACLFilledChecklist use cases?

Reply:
Currently the ACLFilledChecklist has two ways to get server Certificate:
1) Check if the ACLFilledChecklist::serverCert is set 
2) Try getting it from ConnStateData::sslServerBump structure if this is available. 

The (2) is available for peek-and-splice modes.  It is not available when we are  bumping on step1 (client-first bumping mode), we are not building the ConnStateData::sslServerBump object in this case because in this mode we do not need server side TLS connection info (to build the client-to-squid TLS connection). Probably this is a bug we should fix (but I believe not with this PR).

We are setting directly the ACLFilledChecklist::serverCert when we are running server certificate related ACLs (sslproxy_cert_error acl) inside ssl_verify_cb (ssl/support.cc file) and (with this patch) inside Security::PeerConnector to check errors returned from validator. This is an exception, because normally retrieved from the related sslServerBump object.

The sslServerBump is not build yet while we are checking serverCertificate. Moreover 
we need this exception to handle cases the PeerConnector is used to connect to a CachePeer or connect to an ICAPS server.  Maybe we need to add similar set in other places for example in ACLs for ICAP logging.

Reply:
> Alex: Why is this logic not located inside ACLFilledChecklist instead of burdening individual ACLFilledChecklist use cases?

> Christos: Here is how the code works right now: ...

I do not think your description of existing problems answers my question. Yes, ACLFilledChecklist has two messy ways of getting the server certificate. Clearly, those two ways are not sufficient and/or not adequate -- they led to the bug this PR is fixing. Why not add a third way so that the ACLFilledChecklist user does not need to remember all of this?

If, after adding that third way (via `.ssl` field in fd_table, I presume), you discover that one of the old ways (e.g., via ACLFilledChecklist::serverCert) can be removed, great!

Reply:
> If, after adding that third way (via .ssl field in fd_table, I presume), you discover that one of the old ways (e.g., via ACLFilledChecklist::serverCert) can be removed, great!

Not easy for me to find such way. We do not have access to squid-to-tls-server fd inside ACLFilledChecklist nor to the related ssl object from what I can see (we have such access only in the case of pinned connections but this is equivalent to Ssl::ServerBump object method). 

Maybe we can add a new member ACLFilledChecklist:server_fd but still we should give the server_fd or from ACLFilledChecklist constructor, or set directly the member.

I must note that the server certificate still will not be able to be retrieved from the related SSL object in the case of ssl_verify_cb callback because the server certificate is not yet attached to SSL object.

Reply:
> We do not have access to squid-to-tls-server fd inside ACLFilledChecklist

1. ACLFilledChecklist has access to ALE (.al).
2. ALE has access to HierarchyLogEntry (.hier).
3. HierarchyLogEntry has access to server Connection (.tcpServer).
4. Connection has access to the connection FD (.fd).

Now, looking _specifically_ at each location in the current (including this PR) code that is using squid-to-tls-server FD to find the certificate:

* If all of the above items are set, then ACLFilledChecklist can use that information to calculate the certificate pointer without callers help. ACLFilledChecklist should be enhanced. The callers should be simplified.
* Otherwise, we should fix the caller (or some earlier code) so that all of the above fields are set. Those fields are needed for many other things, of course.

Needless to say, the above fields should always be set when the corresponding information exists, but this PR does not need to fix all the places where we forget to set them. We must fix the places that trigger the bug that the customer suffers from. Other places should be fixed in this PR only if such fixes are easy/safe/etc.

> I must note that the server certificate still will not be able to be retrieved from the related SSL object in the case of ssl_verify_cb callback because the server certificate is not yet attached to SSL object.

Agreed, ssl_verify_cb is a special case! It justifies .serverCert existence.

However, it would be good to fix other, non-exceptional cases. The primary point here is not so much to get rid of serverCert as such -- that would be great but, based on your analysis, impossible -- but to reduce serverCert usage to truly exceptional/unusual case(s), so that the remaining ordinary code does not need to worry about remembering to set serverCert.

Reply:
Handled with the commit f05e1341.
Some comments about:
 - Inside Security::PeerConnector class the ACLFilledCheckList::serverCert is set manually. This is to handle the IcapPeerConnector case.
 - The server_cert_fingerprint does not work correctly when it is used with the icap_log lines. In this case uses the remote HTTP server certificate. Not easy to fix it without many changes in src/adaptation/icap/Xaction.cc file, so I let this bug unfixed,. 
- The ACLServerCertificateStrategy::match try three methods to retrieve server certificate. Alternatively we can do it inside ACLFilledChecklist constructor, but in this case we are going to set ACLFilledChecklist::serverCert also in cases we are not interested.

Reply:
Given f05e134, can we remove the manual `check->serverCert` in this Security::PeerConnector::sslCrtvdCheckForErrors() method, possibly making all PR changes in this method unnecessary? My other change request (about documenting the `serverCert` field) assumes that the answer is "yes".

Reply:
We cannot remove the manual set. This is because the Security::PeerConnector class is used by the IcapPeerConnector. If we use ALE or ServerBump to retrieve serverCertificate for IcapPeerConnector case,  will retrieve the remote Http Server certificate not the ICAP server certificate.
At the same time the ssl_verify_cb will use the ICAP server certificate.

Reply:
Clarified in https://github.com/measurement-factory/squid/pull/49/commits/4746043668443b213978c61e2301f8ecab5e5c27